pkg/lib/cockpit-components-dialog.jsx

```diff
@@ -209,14 +129,92 @@
+ *  - footerProps:
+ *      - cancel_clicked optional
+ *          Callback called when the dialog is canceled
  */
 class Dialog extends React.Component {
+    constructor(props) {
+        super(props);
+        this.state = {
+            action_in_progress: false,
+            action_in_progress_promise: null,
+            action_progress_message: '',
+            action_progress_cancel: null,
+            action_canceled: false,
+        };
+        this.update_progress = this.update_progress.bind(this);
+        this.cancel_click = this.cancel_click.bind(this);
+    }
+
+    update_progress(msg, cancel) {
+        this.setState({ action_progress_message: msg, action_progress_cancel: cancel });
+    }
+
     componentDidMount() {
         // if we used a button to open this, make sure it's not focused anymore
         if (document.activeElement)
             document.activeElement.blur();
     }
 
+    action_click(handler, caption, e) {
+        this.setState({
+            action_progress_message: '',
+            action_in_progress: true,
+            action_caption_in_progress: caption,
+            action_canceled: false,
+        });
+
+        const p = handler(this.update_progress)
+                .then(() => {
+                    this.props.set_error(null);
+                    this.setState({ action_in_progress: false });
+                    if (this.props.footerProps.dialog_done)
+                        this.props.footerProps.dialog_done(true);
+                })
+                .catch(error => {
+                    if (this.state.action_canceled) {
+                        if (this.props.footerProps.dialog_done)
+                            this.props.dialog_done(false);
+                    } else {
+                        this.props.set_error(error);
+                        this.setState({ action_in_progress: false });
+                    }
+                    /* Always log global dialog errors for easier debugging */
+                    if (error)
+                        console.warn(error.message || error.toString());
+                });
+
+        if (p.progress)
+            p.progress(this.update_progress);
+
+        this.setState({ action_in_progress_promise: p });
+        if (e)
+            e.stopPropagation();
+    }
+
+    cancel_click(e) {
+        this.setState({ action_canceled: true });
+
+        if (this.props.footerProps.cancel_clicked)
+            this.props.footerProps.cancel_clicked();
+
+        // an action might be in progress, let that handler decide what to do if they added a cancel function
+        if (this.state.action_in_progress && this.state.action_progress_cancel) {
+            this.state.action_progress_cancel();
+            return;
+        }
+        if (this.state.action_in_progress && 'cancel' in this.state.action_in_progress_promise) {
+            this.state.action_in_progress_promise.cancel();
+            return;
```

Comment:
These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19200-20230821-124816-2fdc0a03-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)